[ROCm] Fixes and improvements to CUDA->HIP flag conversion for CPP extensions

naromero77amd opened the pull request:
Fixes https://github.com/ROCm/hip/issues/3764.

Fixes and improvements to CUDA->HIP flag conversion for CPP extensions

- Log flag conversion for debugging purposes.
- Fix cases where it should not touch the -I flags or cases where CUDA appears more than once by replacing only the first instance.
- Fix case where nvcc key may not exist
- Fix case where hipify should ignore flag values and only touch the flag itself


cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

naromero77amd committed:
[ROCm] simple hipify of nvcc flags for CUDAExtension

Fixes https://github.com/ROCm/hip/issues/3764.

A user convenience.

Qubitium commented on torch/utils/cpp_extension.py:
```diff
@@ -1045,6 +1048,12 @@ def _add_compile_flag(self, extension, flag):
+        # Simple hipify, map CUDA->HIP
+        if isinstance(extension.extra_compile_args, dict):
+            extension.extra_compile_args['nvcc'] = [
+                flag.replace("CUDA", "HIP") for flag in extension.extra_compile_args['nvcc']]
```
I think it would be good to have log/print visibility of these hipify actions: log, print to console, which flags where modified. It would lend to better understanding on what hipify is doing and also track problems if these goes awry.

Qubitium replied:
@naromero77amd  I have opend a PR that merges the log changes and changes that prevent breaking existing flags by targetting in more detail when to do replacement vs existing replace all no matter what approach. 

https://github.com/ROCm/pytorch/pull/1964

jeffdaily replied:
merged

Qubitium committed:
reduce CUDA to HIP flag hipify breaking existing/valid flags (#1964)

I made improvements to the exising PR to that fixes CUDA to HIP flag
conversion.

* Log: We should log this behavior since this can break hipcc down the
line for all we know
* Log: Make it transparent to user what hipify actions are doing so devs
are not left in the dark.
* Fix cases where it should not touch the `-I` flags or case where CUDA
appear more than once by replacing only the first instance.
* Fix case where `nvcc` key may not exist 
* Fix case where hipify should ignore flag values and only touch the
flag itself

@naromero77amd 

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport
@dllehr-amd @jataylo @hongxiayang

Passes the following unit test

```py
import unittest
import sys
from unittest.mock import Mock

class TestHipifyCompileFlags(unittest.TestCase):
    def test_hipify_compile_flags(self):

        class DummyClass:
            # Simple hipify, replace the first occurrence of CUDA with HIP
            # in flags starting with "-" and containing "CUDA", but exclude -I flags
            def _hipify_compile_flags(self, extension):
                if isinstance(extension.extra_compile_args, dict) and 'nvcc' in extension.extra_compile_args:
                    modified_flags = []
                    for flag in extension.extra_compile_args['nvcc']:
                        if flag.startswith("-") and "CUDA" in flag and not flag.startswith("-I"):
                            # check/split flag into flag and value
                            parts = flag.split("=", 1)
                            if len(parts) == 2:
                                flag_part, value_part = parts
                                # replace fist instance of "CUDA" with "HIP" only in the flag and not flag value
                                modified_flag_part = flag_part.replace("CUDA", "HIP", 1)
                                modified_flag = f"{modified_flag_part}={value_part}"
                            else:
                                # replace fist instance of "CUDA" with "HIP" in flag
                                modified_flag = flag.replace("CUDA", "HIP", 1)
                            modified_flags.append(modified_flag)
                            print(f'Modified flag: {flag} -> {modified_flag}', file=sys.stderr)
                        else:
                            modified_flags.append(flag)
                    extension.extra_compile_args['nvcc'] = modified_flags

        dummy = DummyClass()
        extension = Mock()

        # (tested flag, expected flag)
        test_cases = [
            ('-CUDA', '-HIP'),
            ('--CUDA', '--HIP'),
            ('--CUDA_CUDA', '--HIP_CUDA'),  # only first CUDA instance
            ('-U__CUDA', '-U__HIP'),
            ('-D__CUDA', '-D__HIP'),
            ('-I/usr/local/CUDA', '-I/usr/local/CUDA'),  # skip includes
            ('-Dmyflag=CUDA', '-Dmyflag=CUDA'),  # should not be touched
        ]

        tested_flags = [test_case[0] for test_case in test_cases]
        expected_flags = [test_case[1] for test_case in test_cases]

        # test `nvcc` key exists
        extension.extra_compile_args = {'nvcc': tested_flags}
        dummy._hipify_compile_flags(extension)
        self.assertEqual(extension.extra_compile_args['nvcc'], expected_flags)

        # test `nvcc` key does not exists
        test_cases = {'abcd': ["-DTest", "--TEST"]}
        extension.extra_compile_args = test_cases
        dummy._hipify_compile_flags(extension)
        # Assert the result
        self.assertEqual(extension.extra_compile_args['abcd'], test_cases['abcd'])

if __name__ == '__main__':
    unittest.main()
```

Qubitium commented:
@naromero77amd @jeffdaily  I think it would be good to update the PR notes and change 

```
A user convenience.
```

To something more accurate for PR history. Because this message makes it sound like it's user's fault. And devs would say this is actually a missing piece of the hipify process and it should've been doing all along. Hipify, the process, overrides everything cuda, including compilers, so it makes total sense from developer's point of view that everything includes compiler flags.

naromero77amd commented:
@Qubitium Updated description. As a co-author you should also be able to make edits to the title and/or description.

> @naromero77amd @jeffdaily I think it would be good to update the PR notes and change
> 
> ```
> A user convenience.
> ```
> 
> To something more accurate for PR history. Because this message makes it sound like it's user's fault. And devs would say this is actually a missing piece of the hipify process and it should've been doing all along. Hipify, the process, overrides everything cuda, including compilers, so it makes total sense from developer's point of view that everything includes compiler flags.

naromero77amd commented:
@pytorchbot merge

jeffdaily commented:
@pytorchbot cherry-pick --onto release/2.7 -c critical
